# Capstone Project : Problem Statement
Businesses try their best to resolve the complaints that they receive.
However, it might not always be possible to appease every customer.
Unhappy consumers might raise follow-up questions/complaints about the
resolutions provided, and this is detrimental to the business as it points to
systemic failures in the Customer Support division and could lead to poor
brand image. Disputed complaints which are being/have been resolved
could be a critical dataset to derive essential learnings for any business.
You have to predict whether a complaint resolution will be accepted or
rejected by a consumer.

# What we do in this notebook

	Data Wrangling and exploration

    Looking into given input dataset and details. 


# Pre-requisite:
    Ensure datafolder and required csv datafile is in place before running the code. 

In [1]:
#Modify the directory & File name based on your local paths & file name
dataFolder = "C:/Users/shaaruns.CORPDOM/Documents/BITS_Assignments/Module-7/Gr20_Cap4_Sourcecode/Grp20_Cohert3_Final_Submission"
datafile = "Input_Dataset/BaseData_complain_Hist.csv"

In [2]:
#import librararies
#Pandas provide powerful and easy-to-use data structures, as well as the means to quickly perform operations on these structures
import pandas as pd
#NumPy library provides objects for multi-dimensional arrays, Numpy is memory efficient compared to panda
import numpy as np
#Matplotlib is a plotting library for the Python programming language and its numerical mathematics extension NumPy. 
#It provides an object-oriented API for embedding plots into applications using general-purpose GUI toolkits
import matplotlib.pyplot as plt
import matplotlib
#Seaborn is a Python data visualization library based on matplotlib. 
#It provides a high-level interface for drawing attractive and informative statistical graphics
import seaborn as sns
#check the versions to avoid compatibility issues
print('pandas  : {}'.format(pd.__version__)) 
print ('numpy   : {}'.format(np.__version__))
print ('matplotlib : {}'.format(matplotlib.__version__))
print ('seaborn : {}'.format(sns.__version__))


pandas  : 0.25.1
numpy   : 1.19.4
matplotlib : 3.1.1
seaborn : 0.9.0


In [3]:
#read the data set in dataframe
dataset_Raw = pd.read_csv(dataFolder+'/'+ datafile)

In [4]:
# sample of data (first 20 rows)
dataset_Raw.head(20)

Date received                  Product  \
0     2015-04-26  Bank account or service   
1     2015-01-30          Debt collection   
2     2014-01-10              Credit card   
3     2015-12-07              Credit card   
4     2012-04-05  Bank account or service   
5     2015-07-22            Consumer Loan   
6     2012-01-06                 Mortgage   
7     2015-04-01          Debt collection   
8     2015-08-04          Debt collection   
9     2012-02-29                 Mortgage   
10    2014-01-21                 Mortgage   
11    2014-04-03          Debt collection   
12    2015-03-08         Credit reporting   
13    2015-07-27                 Mortgage   
14    2015-06-09                 Mortgage   
15    2014-04-06                 Mortgage   
16    2015-08-13             Student loan   
17    2012-08-21              Credit card   
18    2012-04-26                 Mortgage   
19    2015-09-01                 Mortgage   

                               Sub-product  \
0               Other bank product/service   
1                                  Medical   
2                                      NaN   
3                                      NaN   
4                         Checking account   
5                             Vehicle loan   
6              Conventional fixed mortgage   
7                              Credit card   
8                              Credit card   
9                           Other mortgage   
10  Conventional adjustable mortgage (ARM)   
11                             Credit card   
12                                     NaN   
13                             VA mortgage   
14                          Other mortgage   
15  Conventional adjustable mortgage (ARM)   
16                Non-federal student loan   
17                                     NaN   
18             Conventional fixed mortgage   
19                          Other mortgage   

                                       Issue  \
0                   Deposits and withdrawals   
1            Disclosure verification of debt   
2                       APR or interest rate   
3                             Payoff process   
4                   Deposits and withdrawals   
5               Taking out the loan or lease   
6   Application, originator, mortgage broker   
7            Disclosure verification of debt   
8                      Communication tactics   
9   Loan modification,collection,foreclosure   
10  Loan modification,collection,foreclosure   
11       Improper contact or sharing of info   
12  Credit reporting company's investigation   
13              Settlement process and costs   
14  Loan modification,collection,foreclosure   
15  Loan modification,collection,foreclosure   
16                       Can't repay my loan   
17             Credit line increase/decrease   
18  Loan modification,collection,foreclosure   
19  Loan servicing, payments, escrow account   

                                   Sub-issue  \
0                                        NaN   
1       Not given enough info to verify debt   
2                                        NaN   
3                                        NaN   
4                                        NaN   
5                                        NaN   
6                                        NaN   
7       Not given enough info to verify debt   
8                 Frequent or repeated calls   
9                                        NaN   
10                                       NaN   
11     Talked to a third party about my debt   
12  No notice of investigation status/result   
13                                       NaN   
14                                       NaN   
15                                       NaN   
16        Can't decrease my monthly payments   
17                                       NaN   
18                                       NaN   
19                                       NaN   

                         Consumer complaint narrative  \
0   RE : XXXX X

In [5]:
# dataset information
dataset_Raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274019 entries, 0 to 274018
Data columns (total 18 columns):
Date received                   274019 non-null object
Product                         274019 non-null object
Sub-product                     195350 non-null object
Issue                           274019 non-null object
Sub-issue                       105537 non-null object
Consumer complaint narrative    43766 non-null object
Company public response         51783 non-null object
Company                         274019 non-null object
State                           271827 non-null object
ZIP code                        271820 non-null object
Tags                            38506 non-null object
Consumer consent provided?      78285 non-null object
Submitted via                   274019 non-null object
Date sent to company            274019 non-null object
Company response to consumer    274019 non-null object
Timely response?                274019 non-null object
Consumer disp

In [6]:
#describe the data but it is displaying only integer columns.
dataset_Raw.describe()

Complaint ID
count  2.740190e+05
mean   1.045274e+06
std    5.963397e+05
min    5.000000e+00
25%    5.320385e+05
50%    1.066809e+06
75%    1.563406e+06
max    2.126221e+06

The dataset has 18 columns 
dtypes: int64(1), object(17)
274019 entries, 0 to 274018

In [7]:
# columns available in the dataset
dataset_Raw.columns

Index(['Date received', 'Product', 'Sub-product', 'Issue', 'Sub-issue',
       'Consumer complaint narrative', 'Company public response', 'Company',
       'State', 'ZIP code', 'Tags', 'Consumer consent provided?',
       'Submitted via', 'Date sent to company', 'Company response to consumer',
       'Timely response?', 'Consumer disputed?', 'Complaint ID'],
      dtype='object')

In [8]:
#shape of the data
dataset_Raw.shape

(274019, 18)

##### Funtion to identify missing values in column and the pecentage of missing

In [9]:
#% of missing values in each column
def find_missing_values_func(df):
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Selected dataframe has " + str(df.shape[1]) + " columns.\n"      
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        return mis_val_table_ren_columns

In [10]:
#Call the find_missing_values_func(df)
columnsWiseMissingValue = find_missing_values_func(dataset_Raw) 
print(columnsWiseMissingValue)

Selected dataframe has 18 columns.
There are 8 columns that have missing values.
                              Missing Values  % of Total Values
Tags                                  235513               85.9
Consumer complaint narrative          230253               84.0
Company public response               222236               81.1
Consumer consent provided?            195734               71.4
Sub-issue                             168482               61.5
Sub-product                            78669               28.7
ZIP code                                2199                0.8
State                                   2192                0.8


Tags, Consumer complaint narrative, Company public response,Consumer consent provided?  and Sub-issue columns most of the data is missing. More than 60% of data is missing.Sub-product is missing around 28%


Issue: Issue depends on both product and sub-product. Different products and sub-products have different issues.
Sub-Issue: Sub-issue depends on product, sub-product and issue. Hence, it is likely that it will have more NaN values.
Customer complaint: We have large number of nan values in this column. The best possible explaination of this is that customer has not given or retracted his consent on making complaint public( CFCB provides this option).
Sub-product: Every product doesn't have a subproduct for ex. vehicle loan . So naturally we will have some NaN values in sub-product.


# Check duplicate rows and delete 
### Check duplicate Complaint ID and delete last and keep first    ( to be checked)

In [11]:
duplicateRowsDF = dataset_Raw[dataset_Raw.duplicated()]
print("Duplicate Rows except first occurrence based on all columns are :")
print(duplicateRowsDF)

duplicateRowsDF = dataset_Raw[dataset_Raw.duplicated('Complaint ID')]
print("Duplicate Rows except first occurrence based on Complaint ID column are :")
print(duplicateRowsDF)

Duplicate Rows except first occurrence based on all columns are :
Empty DataFrame
Columns: [Date received, Product, Sub-product, Issue, Sub-issue, Consumer complaint narrative, Company public response, Company, State, ZIP code, Tags, Consumer consent provided?, Submitted via, Date sent to company, Company response to consumer, Timely response?, Consumer disputed?, Complaint ID]
Index: []
Duplicate Rows except first occurrence based on Complaint ID column are :
Empty DataFrame
Columns: [Date received, Product, Sub-product, Issue, Sub-issue, Consumer complaint narrative, Company public response, Company, State, ZIP code, Tags, Consumer consent provided?, Submitted via, Date sent to company, Company response to consumer, Timely response?, Consumer disputed?, Complaint ID]
Index: []


Inference: No duplicates based on entire row and Compliant ID

# Compute some derived columns from data files to get some numeric data

### Extracting Date, Month and Year from Date Column¶


In [12]:
#### Using Date received 
dataset_Raw['Year_Received']  = pd.DatetimeIndex(dataset_Raw['Date received']).year
dataset_Raw['Month_Received'] = pd.DatetimeIndex(dataset_Raw['Date received']).month
dataset_Raw['Day_Received']   = pd.DatetimeIndex(dataset_Raw['Date received']).day

#### Date sent to company 
dataset_Raw['Year_Sent']  = pd.DatetimeIndex(dataset_Raw['Date sent to company']).year
dataset_Raw['Month_Sent'] = pd.DatetimeIndex(dataset_Raw['Date sent to company']).month
dataset_Raw['Day_Sent']   = pd.DatetimeIndex(dataset_Raw['Date sent to company']).day


# Find time in days to response by the company¶

In [13]:
dataset_Raw['Date received'] = pd.to_datetime(dataset_Raw['Date received'])
dataset_Raw['Date sent to company'] = pd.to_datetime(dataset_Raw['Date sent to company'])

dataset_Raw['Days held'] = dataset_Raw['Date sent to company'] - dataset_Raw['Date received']

In [14]:
dataset_Raw.dtypes

Date received                    datetime64[ns]
Product                                  object
Sub-product                              object
Issue                                    object
Sub-issue                                object
Consumer complaint narrative             object
Company public response                  object
Company                                  object
State                                    object
ZIP code                                 object
Tags                                     object
Consumer consent provided?               object
Submitted via                            object
Date sent to company             datetime64[ns]
Company response to consumer             object
Timely response?                         object
Consumer disputed?                       object
Complaint ID                              int64
Year_Received                             int64
Month_Received                            int64
Day_Received                            

In [15]:
#Object type of Days held is timedelta64[ns]. So convert it to integer
dataset_Raw['Days held'] = dataset_Raw['Days held'].astype('timedelta64[D]').astype(int)

In [16]:
print(dataset_Raw['Days held'].dtypes)

int32


In [17]:
# we are not going to drop any columns now because we may need to perform further analysis to decide on it.
#dataset_Raw = dataset_Raw.drop(['Date received', 'Date sent to company'], axis = 1)

# Find weeks-number based on Days received

In [18]:
dataset_Raw.Day_Received.unique()

array([26, 30, 10,  7,  5, 22,  6,  1,  4, 29, 21,  3,  8, 27,  9, 13, 24,
       14,  2, 18, 15, 11, 25, 12, 23, 28, 20, 19, 17, 16, 31],
      dtype=int64)

In [19]:
week_train = []  # Define a null list 
for i in dataset_Raw['Day_Received']:
    if i < 8:
        week_train.append(1)
    elif i >= 8 and i < 16:
        week_train.append(2)
    elif i >=16 and i < 22:
        week_train.append(3)
    else:
        week_train.append(4)
dataset_Raw['Week_Received'] = week_train

In [20]:
dataset_Raw.head()

Date received                  Product                 Sub-product  \
0    2015-04-26  Bank account or service  Other bank product/service   
1    2015-01-30          Debt collection                     Medical   
2    2014-01-10              Credit card                         NaN   
3    2015-12-07              Credit card                         NaN   
4    2012-04-05  Bank account or service            Checking account   

                             Issue                             Sub-issue  \
0         Deposits and withdrawals                                   NaN   
1  Disclosure verification of debt  Not given enough info to verify debt   
2             APR or interest rate                                   NaN   
3                   Payoff process                                   NaN   
4         Deposits and withdrawals                                   NaN   

                        Consumer complaint narrative Company public response  \
0  RE : XXXX XXXX XXXX-PRIVILEGED AND CONFIDENTIA...                     NaN   
1                                                NaN                     NaN   
2                                                NaN                     NaN   
3                                                NaN                     NaN   
4                                                NaN                     NaN   

                                     Company State ZIP code  ...  \
0                      Wells Fargo & Company    GA    319XX  ...   
1  HCFS Health Care Financial Services, Inc.    CA    90036  ...   
2                 TD Bank US Holding Company    FL    33180  ...   
3                                Capital One    IL    60610  ...   
4             Citizens Financial Group, Inc.    PA    18337  ...   

  Consumer disputed? Complaint ID Year_Received Month_Received Day_Received  \
0                Yes      1347613          2015              4           26   
1                Yes      1218613          2015              1           30   
2                Yes       663281          2014              1           10   
3                Yes      1685130          2015             12            7   
4                Yes        44225          2012              4            5   

  Year_Sent Month_Sent  Day_Sent  Days held  Week_Received  
0      2015          4        26          0              4  
1      2015          1        30          0              4  
2      2014          1        15          5              2  
3      2015         12         7          0              1  
4      2012          4         5          0              1  

[5 rows x 26 columns]

In [21]:
dataset_Raw.columns

Index(['Date received', 'Product', 'Sub-product', 'Issue', 'Sub-issue',
       'Consumer complaint narrative', 'Company public response', 'Company',
       'State', 'ZIP code', 'Tags', 'Consumer consent provided?',
       'Submitted via', 'Date sent to company', 'Company response to consumer',
       'Timely response?', 'Consumer disputed?', 'Complaint ID',
       'Year_Received', 'Month_Received', 'Day_Received', 'Year_Sent',
       'Month_Sent', 'Day_Sent', 'Days held', 'Week_Received'],
      dtype='object')

###### Find the complaint received Weekday

In [22]:
import datetime as dt
#tags[str(dataset_Raw['Date received'].isoweekday())]
dataset_Raw['weekday'] = dataset_Raw[['Date received']].apply(lambda x: dt.datetime.strftime(x['Date received'], '%A'), axis=1)


In [23]:
dataset_Raw

Date received                  Product                  Sub-product  \
0         2015-04-26  Bank account or service   Other bank product/service   
1         2015-01-30          Debt collection                      Medical   
2         2014-01-10              Credit card                          NaN   
3         2015-12-07              Credit card                          NaN   
4         2012-04-05  Bank account or service             Checking account   
...              ...                      ...                          ...   
274014    2016-01-28          Debt collection                         Auto   
274015    2013-09-20  Bank account or service              Savings account   
274016    2015-03-02                 Mortgage  Conventional fixed mortgage   
274017    2012-08-01  Bank account or service             Checking account   
274018    2012-09-14  Bank account or service             Checking account   

                                           Issue  \
0                       Deposits and withdrawals   
1                Disclosure verification of debt   
2                           APR or interest rate   
3                                 Payoff process   
4                       Deposits and withdrawals   
...                                          ...   
274014           Disclosure verification of debt   
274015                  Deposits and withdrawals   
274016  Loan servicing, payments, escrow account   
274017                 Using a debit or ATM card   
274018   Account opening, closing, or management   

                                   Sub-issue  \
0                                        NaN   
1       Not given enough info to verify debt   
2                                        NaN   
3                                        NaN   
4                                        NaN   
...                                      ...   
274014  Not given enough info to verify debt   
274015                                   NaN   
274016                                   NaN   
274017                                   NaN   
274018                                   NaN   

                             Consumer complaint narrative  \
0       RE : XXXX XXXX XXXX-PRIVILEGED AND CONFIDENTIA...   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
274014  Equifax continues to report a collection accou...   
274015                                                NaN   
274016                                                NaN   
274017                                                NaN   
274018                                                NaN   

       Company public response                                    Company  \
0                          NaN                      Wells Fargo & Company   
1                          NaN  HCFS Health Care Financial Services, Inc.   
2                          NaN                 TD Bank US Holding Company   
3                          NaN                                Capital One   
4                          NaN             Citizens Financial Group, Inc.   
...                        ...                                        ...   
274014                     NaN                                    Equifax   
274015                     NaN                              PNC Bank N.A.   
274016                     NaN                       JPMorgan Chase & Co.   
274017                     NaN                       JPMorgan Chase & Co.   
274018                     NaN                      Wells Fargo & Company   

       State ZIP code  ... Complaint ID Year_Received Month_Received  \
0         GA    319XX  ...      1347613          2015              4   
1         CA    90036  ...      1218613          2015             

In [24]:
dataset_Raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274019 entries, 0 to 274018
Data columns (total 27 columns):
Date received                   274019 non-null datetime64[ns]
Product                         274019 non-null object
Sub-product                     195350 non-null object
Issue                           274019 non-null object
Sub-issue                       105537 non-null object
Consumer complaint narrative    43766 non-null object
Company public response         51783 non-null object
Company                         274019 non-null object
State                           271827 non-null object
ZIP code                        271820 non-null object
Tags                            38506 non-null object
Consumer consent provided?      78285 non-null object
Submitted via                   274019 non-null object
Date sent to company            274019 non-null datetime64[ns]
Company response to consumer    274019 non-null object
Timely response?                274019 non-null obje

 ## Find the pairwise correlation of all columns in the dataframe

In [25]:
#The measure is best used in variables that demonstrate a linear relationship between each other.
dataset_Raw.corr()

Complaint ID  Year_Received  Month_Received  Day_Received  \
Complaint ID        1.000000       0.972538       -0.010833     -0.001111   
Year_Received       0.972538       1.000000       -0.221762     -0.014937   
Month_Received     -0.010833      -0.221762        1.000000     -0.011659   
Day_Received       -0.001111      -0.014937       -0.011659      1.000000   
Year_Sent           0.972117       0.997280       -0.210906     -0.008889   
Month_Sent         -0.019957      -0.219679        0.944362     -0.006575   
Day_Sent           -0.012396      -0.020891       -0.010320      0.635347   
Days held          -0.112025      -0.101857       -0.047438      0.011290   
Week_Received       0.002952      -0.010521       -0.010045      0.964811   

                Year_Sent  Month_Sent  Day_Sent  Days held  Week_Received  
Complaint ID     0.972117   -0.019957 -0.012396  -0.112025       0.002952  
Year_Received    0.997280   -0.219679 -0.020891  -0.101857      -0.010521  
Month_Received  -0.210906    0.944362 -0.010320  -0.047438      -0.010045  
Day_Received    -0.008889   -0.006575  0.635347   0.011290       0.964811  
Year_Sent        1.000000   -0.229945 -0.026197  -0.085019      -0.005427  
Month_Sent      -0.229945    1.000000 -0.016085   0.038062      -0.004580  
Day_Sent        -0.026197   -0.016085  1.000000   0.010423       0.645106  
Days held       -0.085019    0.038062  0.010423   1.000000       0.010115  
Week_Received   -0.005427   -0.004580  0.645106   0.010115       1.000000

# Save first phase of cleaned up file for future analysis

In [26]:
## Save first phase of cleaned up file

dataset_Raw.to_csv(dataFolder+"/Output_CSV/Dataset_Step1_DataWrangling.csv",index=False)#File name change